src/view/xwayland.cpp

```diff
@@ -198,23 +198,51 @@ class wayfire_xwayland_view_base : public wf::wlr_view_t
         /* Wayfire positions views relative to their output, but Xwayland
          * windows have a global positioning. So, we need to make sure that we
          * always transform between output-local coordinates and global
-         * coordinates */
-        if (get_output())
-        {
-            auto current_workspace =
-                get_output()->workspace->get_current_workspace();
-            auto wsize = get_output()->workspace->get_workspace_grid_size();
-            auto og    = get_output()->get_layout_geometry();
+         * coordinates. Additionally, when clients send a configure request
+         * after they have already been mapped, we keep the view on the
+         * workspace where its center point was from last configure, in
+         * case the current workspace is not where the view lives */
+        auto o = get_output();
+        if (o)
+        {
+            auto wsg = o->workspace->get_workarea();
+            auto og  = o->get_layout_geometry();
             configure_geometry.x -= og.x;
             configure_geometry.y -= og.y;
+            auto view = this->self();
+            while (view->parent)
+            {
+                view = view->parent;
+            }
+
+            auto vg = view->get_wm_geometry();
+            auto current_workspace = o->workspace->get_current_workspace();
+            if (vg.width && vg.height)
+            {
+                vg.x += current_workspace.x * og.width;
+                vg.y += current_workspace.y * og.height;
+                auto center = wf::point_t{vg.x + vg.width / 2, vg.y + vg.height / 2};
+                auto ws     =
+                    wf::point_t{int(center.x / og.width), int(center.y / og.height)};
+                wsg.x += (ws.x - current_workspace.x) * og.width;
+                wsg.y += (ws.y - current_workspace.y) * og.height;
+            }
+
+            if (view_impl->frame)
+            {
```

Comment:
You are looking for get_output_geometry() here. A badly named function, but it gives you the geometry without decorations.

Reply:
I already have the decorationless size, because that's what we compute for the configure geometry. I actually need to know the titlebar height and the border size so that I can clamp the X geometry (recall that's without borders) to the workarea, before the borders are taken into account. Basically I have to shrink the workarea by the decoration sizes on their respective edges. get_output_geometry() will give me an old size anyway.

Reply:
Instead of comparing undecorated geometry to shrunk workspace geometry, you can just as well compare the expanded geometry to the whole workspace (the math works out if you think about it)